### DISCUSSION TITLE
[5기] 모바일 안드로이드 레벨4 바퀴의 재발명 | 2023.09.01

### DISCUSSION THREAD

**Discussion by woowahan-pjs** [woowacourse/retrospective#41] [General]
"[바퀴의 재발명(To reinvent the wheel)](https://en.wikipedia.org/wiki/Reinventing_the_wheel)"이라는 단어를 들어본 적이 있나요? 이 관용구는 **이미 발명되었고 본질적인 결함이 없는 것을 다시 발명하는 것은 가치를 더하지 못하고 시간 낭비**일 뿐이라는 의미로 "**바퀴를 재발명하지 말라**"라고 말하는 데 주로 사용됩니다.
하지만 **때로는 "바퀴를 재발명"하는 것이 필요**할 때도 있습니다. 어떤 상황에서 바퀴를 재발명하는 행위가 필요할까요? **재발명을 할 때 얻을 수 있는 이점과 단점에 대해 생각해 보고 500자 이내로 작성**해 주세요. 내일 수업 시간에 각자가 쓴 내용을 함께 토론해 볼게요.

**Comment by DYGames** [comment]
## 바퀴를 재발명할 때 이점과 단점
### 이점
- 소프트웨어는 여러 추상화된 계층이 겹겹이 쌓여 이루어져 있습니다. 안드로이드와 같은 응용 프로그램 개발은 상당히 고수준의 계층에서 개발이 이루어지며, 저수준 계층의 구현을 알기 어렵습니다.
- 이때 복잡한 기능을 구현하거나 버그가 발생 할 시 문제 해결에 어려움을 겪을 수 있습니다.
- 바퀴를 재발명함은 이 저수준 계층의 구현을 재현함으로 저수준 계층의 구현을 학습할 수 있습니다. 그러므로 고수준 프로그램의 수준을 높이고 버그 발생 가능성을 줄일 수 있습니다.

- 또한 우리가 사용하는 바퀴(라이브러리 등)가 우리가 개발하고자 하는 응용 프로그램에 정확히 적합한 바퀴라고 할 수 없습니다. 이런 경우에 바퀴를 재발명하여 코드 관리, 안전성, 성능 측면에서 이점을 가져갈 수 있습니다.

### 단점
- 개발 속도가 중요한 상황에서 기존의 훌륭한 바퀴를 그대로 사용하면 일정 관리에 도움이 될 수 있지만 이 바퀴를 재발명함은 개발 시간을 지연시키고 불필요한 코드를 추가시킬 수 있습니다.

**Comment by tmdgh1592** [comment]
## 이점

- 바퀴가 고장났을 때, 바퀴를 재발명해보았다면 그 구조를 알고 있기 때문에 문제를 비교적 쉽게 해결할 수 있다.
- 바퀴가 더 이상 판매되지 않아도 상관없다. 내가 만들면 되니까.
- 남이 만든 바퀴의 저작권에 얽매일 필요가 없다.

## 단점

- 이미 발명되어 있는 바퀴보다 더 낫다는 보장이 없다.
- 바퀴를 만들기 위해 본인의 시간을 별도로 투자해야 한다.

**Comment by rhkrwngud445** [comment]
## 이점
- 기존 있던 바퀴가 놓친 부분(결함, 기능)을 보완할 수 있습니다.
- 바퀴 제작의 전문가로 한 걸음 나아갈 수 있습니다. 어떤 바퀴의 종류가 있는지, 바퀴 마다 어떤 장단점이 있는지 면밀히 비교할 수 있는 기회가 되기도 합니다.
- 제가 바퀴를 만들었습니다.

## 단점
- 바퀴가 이미 많이 만들어져 있다면, 그 바퀴를 설계하고 만드는데 비용이 든다는 반증이 됩니다.
- 다 만들었을 때, 바퀴가 없어진 탈 것이 상용화 되어있을 수도 있다.

**Comment by boogi-woogi** [comment]
### 단점

- 바퀴라는 것이 하루아침에 만들어지지 않았다. 많은 시행착오를 거쳐서 현재 사용되고 있는 바퀴의 형태가 만들어졌다. 재발명을 통해 기존의 바퀴보다 성능이 좋을 것이라는 보장이 없을뿐더러 만드는 과정조차 쉽지 않다.

### 이점

- 많은 시행착오를 함께 겪어온 직접 만든 바퀴에 더 많은 애정이 갈 수 있다.
- 자신이 만든 바퀴 내부적으로 문제가 발생했을 때 그 문제에 대해 파악하고 해결해 나가는 것은 만든 사람이 가장 잘 수행할 것이다.
- 만들어져 있는 바퀴를 사용만 해 본 사람과 직접 바퀴를 만들어 본 사람의 시야의 넓이가 확연히 다를 것이다.

**Comment by hyemdooly** [comment]
### 이점
- 기존의 바퀴를 더 잘 이해하게 될 수 있다. -> 기존의 바퀴를 더 잘 쓸 수 있다.
- 우리의 것에 딱 맞는 바퀴를 만들 수 있다.

### 단점
- 기존 바퀴보다 효율적이지 못하거나 좋지 못할 수 있다.
- 바퀴를 만들기 위해 개발 시간이 더 추가된다.

**Comment by ippnsj** [comment]
## 바퀴의 재발명(To reinvent the wheel
### 이점
- 조직에 특화된 바퀴를 만들어 사용 가능하다.
- 바퀴가 조직에 불필요한 기능을 가지고 있지 않다.
- 바퀴를 직접 만들면서 내부적으로 어떻게 동작하는지 배울 수 있다.
- 내부 구조를 잘 알기 때문에 바퀴에 새로운 기능을 추가하는 것이 상대적으로 쉽다.
- 예기치 못한 버그가 발생할 가능성이 상대적으로 적다.

### 단점
- 새로운 바퀴를 만드는데 시간을 투자해야 한다.
- 기존의 바퀴 사용 경험이 있는 사람도 조직에 특화된 바퀴를 사용하는 방법을 배워야 한다.

**Comment by chws0508** [comment]
## 장점 
- 기존 바퀴를 재발명 해봄으로써, 이 기능을 재학습이 가능합니다. 
- 기존 바퀴를 재발명 해보는 과정을 통해 , 앞으로 프로덕트를 더 발전할 수 있는 방향을 다시 생각할 수 있을 것 같습니다.

## 단점
- 기존에 발명했던 바퀴와 성능이 같다고 보장을 할 수가 없습니다.
- 시간을 많이 뺏길 수 있습니다.

**Comment by pingu244** [comment]
## 이점
직접 코드를 치는 것과 가져다 쓰는 것에는 큰 차이가 있다고 생각합니다. 
바퀴를 재발명을 하게 된다면 그 경험과 이해를 바탕으로 개선된 바퀴(네모난 바퀴), 발전된 바퀴(타이어, 도르레) 혹은 바퀴가 아닌 새로운 무언가를 만들어낼 수 있다고 생각합니다.
## 단점
시간 소요가 크다는 것이 가장 큰 단점이라고 생각합니다.
다른 곳에 쓸 시간을 이미 있는 바퀴를 만드는 것에 쓰는 것이니까요.
또 시간을 들여 나만의 바퀴를 만들었다고 해도 그것이 기존의 바퀴보다 낫다고 확신할 수는 없습니다.

**Comment by ki960213** [comment]
# 이점
바퀴를 만들기 위해 바퀴를 만드는 방법을 공부하게 됩니다. 이 과정에서 바퀴에 대한 이해도가 증가할 수 있다고 생각합니다.
바퀴에 대한 이해도가 증가한다면 더 나은 바퀴를 만드는 방법을 떠올릴 수도 있게 됩니다.
바퀴를 더 짧은 시간 안에 만들 수 있게 됩니다.

# 단점
바퀴를 만드는 데 시간이 듭니다.
바퀴를 만드는 데 오랜 시간이 걸렸다면 바퀴에 대해 생각을 많이 했을 것입니다. 그래서 이후에 이동을 원활하게 해주는 부품을 만들려고 생각하면 바퀴가 가장 먼저 떠오르게 될 수 있습니다. 예를 들어 썰매와 수레를 본 적 없는 사람이 눈이 쌓인 곳에서 물건을 옮기기 위해 필요한 도구를 만들려고 할 때 바퀴를 이용한 도구를 만들 수 있다고 생각합니다.

**Comment by SeongHoonC** [comment]
## 이점
1. 바퀴를 만들어본 경험으로 바퀴 공장에 취직할 수 있다.
2. 바퀴를 만드는데 필요한 재료를 알 수 있고 변경할 수 있다.
3. 바퀴를 만드는 방법을 알 수 있다.
4. 바퀴를 만드는데 걸리는 시간 및 난이도를 알 수 있다.
5. 다른 형태의 바퀴가 나올 수 있다. 

## 단점
1. 더 좋은 바퀴를 만드는 것은 어렵다.
2. 시간이 오래 걸린다.
3. 바퀴가 나에 대한 의존성이 생긴다.

**Comment by whk06061** [comment]
# 이점
- 바퀴를 어떻게 만드는지 그 방법과 내부 구조를 알 수 있다.
- 방법을 알게 되면 원하는 모양의 바퀴를 만들 수 있다.

# 단점
- 이미 상용화 되어 있는 바퀴보다 더 나은 성능의 바퀴를 만들기 힘들 수 있다.

**Comment by RightHennessy** [comment]
### 이점

바퀴를 한번 만들어보면서 유사답지(?)가 있는 학습을 할 수 있다.
새로만든 바퀴는 내 입맛대로 커스텀 하기 좋다.

### 단점

이미 발명되었고 본질적인 결함이 없는 바퀴라면, 이러한 바퀴를 재생산하는 과정은 매우 복잡하고 많은 시간이 소요될 것으로 예상된다.

**Comment by sujin9** [comment]
## 이점
- 나에게 더 잘 맞는 바퀴를 만들게 될 수도 있다.
- 바퀴를 만드는 방법을 알게 되고, 바퀴의 작동 원리 및 방법을 이해할 수 있게 된다.

## 단점
- 만드는 데 많은 시간이 소요될 수 있다.
- 바퀴를 필요로 하는 새로운 장비(장치)가 등장했을 때, 이에 맞는 바퀴를 만들어 대응하는 과정이 이미 발명된 바퀴에 비해 오래걸릴 수 있다.
- 이미 발명된 바퀴보다 더 빈번하게 오류가 발생할 가능성이 있다.

**Comment by EmilyCh0** [comment]
### 이점

- 사용하고 있는 바퀴가 최선이 아닐 수 있다. 더 적합한 바퀴를 재발명할 수 있다.
- 바퀴를 재발명하면서 기존에 사용하던 바퀴에 대한 이해도가 더 높아질 수 있다.
- 자신이 개발한 바퀴에 대한 이해도가 높을 것이다. 이후 바퀴에 기능이 추가되거나 변경이 필요할 때 높은 이해도를 바탕으로 빠르게 반영할 수 있다.

### 단점

- 바퀴를 재발명해서 얻는 이점에 비해 리소스가 너무 많이들 수 있다.
- 더 좋은 바퀴를 만든다는 보장이 없다.
- 보편적으로 사용되는 바퀴가 아니기에 다른 사람이 코드를 이해하고 적용하는데도 비용이 든다.

**Comment by s9hn** [comment]
### 이점
- 서비스의 확장 및 변화에 유동적으로 대응할 수 있다.
- 내가 만든 바퀴가 더 나은 바퀴가 되어, 해결하지 못한 문제를 해결해줄수도 있다.

### 단점
- 리소스가 많이 소요된다.

**Comment by no1msh** [comment]
## 이점
- 바퀴를 직접 만들면서 조금더 자기 바퀴가 쓰일 곳에 유리하게 재발명 할 수 있다.
- 바퀴를 만들며 바퀴에 대해 이해하기 쉽다.
- 자신이 기존의 바퀴를 만들었다면, 만들 당시에 바꾸고 싶었던게 있을 수 있는데 그걸 다시 재발명하면서 바꿀 수 있다.

## 단점
- 기존의 바퀴를 쓰면 해결될 문제라고 가정한다면, 당장 시간이 여유롭지 않은 상황에서는 탐탁치 않은 시간 소요라고 볼 수 있다.
- 기존의 바퀴보다 더나은 바퀴를 재발명 할거라는 보장이 없다.

**Comment by Choisehyeon** [comment]
### 이점
- 바퀴의 구조에 대한 이해도가 높아져서 다양한 형태를 가진 바퀴를 만들 수 있다.
- 내가 만든 바퀴이기 때문에 추후 변화가 필요할 때 좀 더 유연하게 대처할 수 있다.

### 단점
- 이미 만들어진 바퀴가 있는데 새로 만드는 것이기 때문에 많은 시간이 소요될 것이다.

**Comment by inseonyun** [comment]
### 이점
+ 바퀴가 어떻게 쓰여지는지 알 수 있다.
+ 바퀴를 좀 더 잘 굴러가게, 성능이 좋게 만들 수 있다.
+ 기존 바퀴의 문제점을 보완해나갈 수 있다.

### 단점
+ 그만한 시간을 필요로 하게 된다.
+ 이전 바퀴와 재발명한 바퀴 + 재발명에 소요된 시간을 비교했을 때 큰 차이가 없다면 재발명 하지 않는게 좋을 수도 있다.

**Comment by otter66** [comment]
## 이점
**변하는 상황에 맞추어 새로운 것에 더 잘 어울리는 바퀴를 만들거나 적용할 수 있는 경험과 지식을 만들 수 있습니다.**
처음으로 만들어진 바퀴는 통나무를 잘라 원판 형태로 만든 것이었다고 합니다. 하지만 압력을 버티지 못해 바큇살이 있는 형태로 변화했고, 고무 소재가 발전하며 이를 자동차에 활용하기도 하였습니다. 또, 같은 자동차용 타이어라고 하더라도 SUV에 어울리는 바퀴, 트럭에 어울리는 바퀴처럼 상황에 알맞은 다양한 종류가 있습니다. 바퀴를 만들며 바퀴의 구조에 대해 이해하고 어떤 상황에 어울리는지 판단할 수 있게 될 것입니다.

## 단점
**시간을 낭비할 수 있고, 오류를 겪을 수 있습니다.**
이미 여러 타이어들이 개발되어 있습니다. 각 타이어들의 특성을 파악할 수 있다면, 새로운 자동차를 만들 때 자동차의 특성에 맞추어 개발되어 있는 타이어를 선택하면 됩니다. 그러나 직접 만들게 된다면 검증되지 않은 성능으로 이전에 겪은 문제를 다시 겪을 수도 있고, 그 과정에서 시간 또한 낭비하게 됩니다.

**Comment by 2chang5** [comment]
# 바퀴의 재발명에 대한 견해


### 바퀴는 왜 사용하는가?

타인이 만들어 놓은 바퀴를 사용하는 이유는 무엇일까? 사실상 시간과 기술력이 무한하다면 개인이 모든것을 만들어서 사용하는것이 가장 이상적일것이다.

외부 의존이 전혀 없기에 어떤 부분이 업데이트 된다고 코드가 고장나는 일도 없을 뿐더러, 코드에 대한 이해도 완벽할테니 더할나위 없이 좋을것이다.

하지만 이것은 현실적으로 불가능하다. 개인의 지식에는 한계가 있고 시간 또한 한계가 있다. 

가벼운 예시를 들어보자 안드로이드 개발을하며 Http통신을 할때 네트워크 지식에 대한 CS에 대해서 모두 파악하고 각각의 기능을 사용하는가? 단순 라이브러리 사용법만 알뿐 Http통신인지 조차 모르는 사람도 존재할것이다. 

이렇듯 잘만든 바퀴는 추상화된 계층을 제공하여 모든 지식을 함양하지 않고도 기능 구현이라는 목표를 달성할 수 있는 생산성을 제공한다.

이는 사실상 원시인들이 우가우가 하며 가내수공업으로 모든것을을 얼기설기 만들어내다 현대사회에 이르러 분업을 통해 각각의 내용물의 퀄리티가 올라간것이나 다름없다.

만약 바퀴를 사용하지 않는다면 모든걸 얼기설기 가내  수공업으로 만드는 우가우가를 자처하는 것이나 다름없다.




### 바퀴의 범위를 어느것으로 잡을것인가?

그렇다고 바퀴가 마냥 좋은것은 아닐것이다. 이 바퀴가 잘돌아가나 항상 지켜보며 윤활유도 발라주고 주기적으로 교체도 해주어야 할것이다.

이미 돌아가고있는 바퀴에 대한 큰 범위를 지정해보면 3가지 정도로 나눌 수 있을것이라 생각한다.



1. 언어적 차원의 지원 api

우선적으로 가장 신뢰하고 사용해야하는 api라고 생각한다. 언어는 아주 똑똑한 사람들에 의해서 만들어지며 대중적인 언어는 그 많은 유저들의 피드백을 통해서 점점 더 발전한 형태이다. 이렇듯 수도 없이 검증된 만큼 바퀴를 새롭게 만들었을때 그것을 뛰어넘기 힘들며 계속해서 언어차원에서 이루어지는 최적화에 대한 이점도 얻어가지 못할테니 정말 필요한 시점이 아니라면 이 바퀴는 그냥 순정으로 쓰는게 좋다고 생각한다.



2. 신뢰도 있는 라이브러리(구글,오랫동안 검증된 라이브러리)

믿을만 하지만 충분히 상황에 맞춰서 바퀴를 재생산 할 수 있다고 생각한다. 

이번 Trip Draw 프로젝트에서 Naver map api를 이용하며 겪은 문제를 통해 왜 큰 기업에서는 외부 의존성을 제거하려고 부단히 노력하는지 알 수 있었다. 이 사례를 살펴보자면 Naver map에서는 내부적으로 Google의 gms play-services-location을 이용하는데  21버전부터 많은 변화가 일어나 interface가 변경되어 Naver map api는 최신 api를 이용하는데 한계가 있었다. 

이처럼 외부의 조건으로 인해 내부 코드가 영향을 받을 수 있으므로 만약 서비스에서 정말 중요한 도메인이라면 언어 차원 이상으로 지원을 받는것보다는 직접 모든 로직을 작성하는것도 고려해봐야한다고 생각한다.

하지만 대부분의 경우 검증된 라이브러리를 사용하는것이 생산성 측면, 혹은 그것을 다룰 수 있는 개발자의 수도 많기 때문에 좋을 뿐더러 그 라이브러리에대한 대체제도 분명 많이 존재할 가능성이 크기에 너무 모든것을 다 만들 필요는 없다고 생각한다.



3. 팀내 코드

굳이 외부의 라이브러리를 제외하고도 팀내 모듈등 많은 바퀴들이 존재할것이다. 이는 우리의 리팩터링, 그리고 바퀴의 재생산에 주요 타겟이라고 생각한다.

개발을 하면서 항상 추후 리팩터링이라는 면죄부아래 더러운 코드를 코드베이스로 밀어넣게된다.(일정 때문에 어쩔수 없지 않는가?) 이렇기 때문에 이미 돌아가고 있더라도 우리는 눈에 불을 키고 고칠건 없는지 새로 만들건 없는지 다시 돌아봐야한다고 생각한다.




## 이점

- 외부와 의존성을 끊어냄으로 최대한 안정적인 프로젝트를 유지할 수 있다.
- 현 도메인에 맞는 상황으로 최적화 할 수 있다.(기능, 성능)

- 바퀴를 너무 잘만들었다면 오픈소스화 시켜서 명성을 드높일 수 있다.

## 단점

- 새로운 개발자가 팀에 입성했을때 우리만의 바퀴를 보고 이해하는데 시간이 오래 걸릴 수 있다.
- 바퀴를 다시 만드는데 드는 비용이 크다.
- 새로 만든 바퀴는 관리대상이 하나 늘어난다는 것과 같다.(관리되지 못한다면 골치덩어리고 외부의 업데이트되는 바퀴를 이용하는것이 좋다.)



p.s 자꾸 바퀴바퀴하니까 바퀴벌레 같아서 화가나네요

**Comment by hyunji1203** [comment]
## 이점

- 기존 바퀴의 구조 및 작동 원리에 대해 학습할 수 있고 이해도도 높일 수 있다.
- 바퀴를 재발명 하게 된다면 기존 바퀴의 기능 뿐만 아니라 넣고 싶은 새로운 기능도 추가할 수 있다

## 단점

- 이미 존재하는 바퀴를 다시 만들려 하기 때문에 불필요한 시간과 노력을 낭비하게 될 수 있다.
- 바퀴를 재 발명하는 과정에서 추가적인 자원 및 비용이 들어갈 수 있으므로 오히려 비 효율적일 수 있다.
- 기존 바퀴보다 더 나은 바퀴를 만들거란 보장이 없다.

**Comment by krrong** [comment]
## 이점
**이해를 높여 목적에 맞게 변형할 수 있다.**
현재의 바퀴는 수많은 사람들이 갈고 닦은 지식의 결정이다. 그리고 그것을 가져와 사용함으로써 바퀴에 투자할 재화들을 다른 곳에 투자할 수 있기에 재발명은 중요하다고 생각하지 않을 수 있다. 그러나 때로는 우리가 추구하는 목적에 따라 바퀴는 변해야 할 수도 있다. 수레를 끄는 바퀴와 화물 트럭을 운반하는 자동차의 바퀴는 목적성은 같지만 크기와 모양이 다르듯 분명한 차이가 존재한다. 그리고 그 간극을 메꾸기 위해서는 바퀴의 재발명은 필수 조건이 될 수 있다.

## 단점
**바퀴를 잘 갈고 닦아 둥글게 만들어야 한다.**
바퀴를 둥글게 만든다는 것은 본래의 목적성을 잃지 않아야 한다는 것을 말하며 나아가 문제없는 기능들을 제공해야 한다. 여기까지 도달하는 과정에서 예상할 수 없는 문제들을 직면하거나 불필요한 작업들이 늘어날 수 있다.

**Comment by re4rk** [comment]
## 이점

1. 바퀴는 이미 많은 검증을 거친 결과물일 것입니다. 재발명의 과정속에서 나오는 결과물과 이미 완성된 것의 차이점을 비교해가면서 어떤 노하우가 담겨있는지 알 수 있을 것입니다.
2. 본질적인 결함이 없다고 생각하였지만 재발명을 통해서 더 좋은 바퀴도 나올 수 있다고 생각하고 있습니다.

## 단점

1. 재발명에 대한 이점이 이를 위한 시간과 돈 투자보다 크다고 볼 수 없습니다.

**Comment by briandr97** [comment]
* 단점
이미 주어진 것을 다시 만들기 때문에 한정된 시간 속에선 매우 비효율적일 수 있다.

* 장점
이미 만들어진 것이더라도 다시 만들어보면서 어떻게 만들어졌는지를 알 수 있다.
어떻게 만들어졌는지를 알면 사용할 때도 더 효율적으로 사용할 수 있으며, 오류가 생기거나 고장났을 때 잘 고칠 수 있다.
또한 수레 바퀴를 다시 만들어서 바퀴 만드는 방법을 익힌다면 수레 바퀴가 아닌 자전거 바퀴가 필요할 때도 유용할 것이다.

**Comment by rhthrhrl0** [comment]
# 장점
- 바퀴가 만들어진 원리를 이해할 수 있다.
- 바퀴가 맘에 들지 않으면, 내가 원하는 바퀴를 만들고 교체할 수 있다

# 단점
- 그 안에 있는 모든 원리를 알기 위해서는 많은 시간 투자를 해야 한다.
